src/domain/usecases/CopyProgramStageDataValuesUseCase.ts

```diff
@@ -0,0 +1,175 @@
+import _ from "lodash";
+import fs from "fs";
+import { Id } from "domain/entities/Base";
+import { DataElement } from "domain/entities/DataElement";
+import { DataElementsRepository } from "domain/repositories/DataElementsRepository";
+import { ProgramEventsRepository } from "domain/repositories/ProgramEventsRepository";
+import { OrgUnitRepository } from "domain/repositories/OrgUnitRepository";
+import { ProgramEvent } from "domain/entities/ProgramEvent";
+import log from "utils/log";
+
+export class CopyProgramStageDataValuesUseCase {
+    constructor(
+        private programEventsRepository: ProgramEventsRepository,
+        private orgUnitRepository: OrgUnitRepository,
+        private dataElementsRepository: DataElementsRepository
+    ) {}
+
+    async execute(options: CopyProgramStageDataValuesOptions): Promise<ProgramEvent[]> {
+        const { programStageId, dataElementIdPairs: idPairs, post, saveReport: reportPath } = options;
+
+        const rootOrgUnit = await this.orgUnitRepository.getRoot();
+        const dataElements = await this.dataElementsRepository.getByIds(idPairs.flat());
+        const dataElementPairs = this.mapDataElements(dataElements, idPairs);
+        const sourceIds = idPairs.map(([sourceId, _]) => sourceId);
+        const targetIds = idPairs.map(([_, targetId]) => targetId);
+
+        checkDataElementTypes(dataElementPairs);
+
+        const allEvents = await this.programEventsRepository.get({
+            programStagesIds: [programStageId],
+            orgUnitsIds: [rootOrgUnit.id],
+            orgUnitMode: "DESCENDANTS",
+        });
+
+        const applicableEvents = allEvents.filter(event =>
+            event.dataValues.some(dv => sourceIds.includes(dv.dataElement.id))
+        );
+
+        checkTargetDataValuesAreEmpty(applicableEvents, targetIds);
+
+        const eventsWithNewDataValues = this.copyEventDataValues(
+            applicableEvents,
+            sourceIds,
+            dataElementPairs
+        );
+
+        if (post) {
+            const result = await this.programEventsRepository.save(eventsWithNewDataValues);
+            if (result.type === "success") log.info(JSON.stringify(result, null, 4));
+            else log.error(JSON.stringify(result, null, 4));
+        } else {
+            const payload = { events: eventsWithNewDataValues };
+            const json = JSON.stringify(payload, null, 4);
+            const now = new Date().toISOString().slice(0, 19).replace(/:/g, "-");
+            const payloadPath = `copy-program-stage-data-values-${now}.json`;
+
+            fs.writeFileSync(payloadPath, json);
+            log.info(`Written payload (${eventsWithNewDataValues.length} events): ${payloadPath}`);
+        }
+
+        if (reportPath) {
+            this.saveReport(reportPath, dataElementPairs, programStageId, eventsWithNewDataValues);
+        }
+
+        return eventsWithNewDataValues;
+    }
```

Comment:
That's too long, specially being the public method of the class. Identify the steps involved and abstract to private methods.